internal/proxyservice/tcp/proxy_service.go

```diff
@@ -113,16 +118,36 @@ func (proxy *proxyService) handleConnection(clientConn net.Conn) error {
 		return fmt.Errorf("failed on retrieve credentials: %s", err)
 	}
 
+	proxy.logger.Infof("new connection on %v.\n", clientConn.LocalAddr())
+
 	targetConn, err = proxy.connector(clientConn, backendCredentials)
 	if err != nil {
 		return fmt.Errorf("failed on connect: %s", err)
 	}
 
-	return <-duplexStream(clientConn, targetConn)
+	proxy.logger.Infof("%v connection to %v opened.\n", clientConn.LocalAddr(), targetConn.RemoteAddr())
+
+	clientErrChan, destErrChan := duplexStream(clientConn, targetConn)
+	select {
+	case sourceErr := <-clientErrChan:
+		if sourceErr == nil {
+			proxy.logger.Infof("%v connection closed by client.\n", clientConn.LocalAddr())
+			return nil
+		}
+		return fmt.Errorf("%v connection failed on streaming from client connection: %s", clientConn.LocalAddr(), sourceErr)
+	case destErr := <- destErrChan:
+		if destErr == nil {
+			proxy.logger.Infof("%v connection closed by target.\n", clientConn.LocalAddr())
```

Comment:
You use the `proxy.logger` a decent amount - maybe we could reduce code if you had `logger := proxy.logger` somewhere higher?